pandas/tests/frame/methods/test_fillna.py

```diff
@@ -525,6 +521,97 @@ def test_fill_corner(self, float_frame, float_string_frame):
         # TODO(wesm): unused?
         result = empty_float.fillna(value=0)  # noqa
 
+    @pytest.mark.parametrize(
+        "expected,fill_value",
+        [
+            (
+                DataFrame(
+                    [[100, 100], [200, 4], [5, 6]], columns=list("AB"), dtype="float64"
+                ),
+                Series([100, 200, 300]),
+            ),
+            (
+                DataFrame(
+                    [[100, 100], [np.nan, 4], [5, 6]],
+                    columns=list("AB"),
+                    dtype="float64",
+                ),
+                {0: 100, 2: 300, 3: 400},
+            ),
+        ],
+    )
+    def test_fillna_column_wise(self, expected, fill_value):
+        # GH 4514
+        df = DataFrame([[np.nan, np.nan], [np.nan, 4], [5, 6]], columns=list("AB"))
+        result = df.fillna(fill_value, axis=1)
+        tm.assert_frame_equal(expected, result)
+
+    def test_fillna_column_wise_downcast(self):
+        # GH 4514
+        df = DataFrame([[np.nan, 2], [3, np.nan], [np.nan, np.nan]], columns=list("AB"))
+        s = Series([100, 200, 300])
+
+        expected = DataFrame(
```

Comment:
what happens if we have a datetime column mixed in here?

Reply:
We end up with object dtype:
``` python
In [5]: df = pd.DataFrame([[np.nan, 2], [3, np.nan], [np.nan, np.nan]], columns=list("AB"))
   ...: s = pd.Series(pd.to_datetime([100, 200, 300], unit="ns"))
   ...: 
   ...: result = df.fillna(s, axis=1, downcast="infer")
   ...: result
Out[5]: 
                               A                              B
0  1970-01-01 00:00:00.000000100                            2.0
1                            3.0  1970-01-01 00:00:00.000000200
2  1970-01-01 00:00:00.000000300  1970-01-01 00:00:00.000000300

In [6]: result.dtypes
Out[6]: 
A    object
B    object
dtype: object
```
``` python
In [8]: df = pd.DataFrame({"A": pd.to_datetime([np.nan, 2, np.nan]), "B": pd.to_datetime([3, np.nan, np.nan])})
   ...: s = pd.Series([100, 200, 300])
   ...: 
   ...: result = df.fillna(s, axis=1, downcast="infer")
   ...: result
Out[8]: 
                               A                              B
0                            100  1970-01-01 00:00:00.000000003
1  1970-01-01 00:00:00.000000002                            200
2                            300                            300

In [9]: result.dtypes
Out[9]: 
A    object
B    object
```

Reply:
is this what we want?